### PR TITLE
installation: Add SLOPE as optional module

### PR DESCRIPTION
**Pull request by FabioLuporini**
Now users can install SLOPE, the library that enables loop tiling (sparse tiling) in PyOP2. 

This is a non-compulsory dependency. Users will be able to install it pretty much in the same way as firedrake-adjoint and slepc, i.e., by passing --slope to the install script

### CODE REVIEW

**Comment by miklos1** [comment]
This change looks fine.

**Comment by dham** [comment]
Looks like it's missing the corresponding Firedrake-zenodo change.

**Comment by FabioLuporini** [comment]
I've now added the proper SLOPE entry to firedrake-zenodo

File: scripts/firedrake-zenodo [hunk 1]
```diff
@@ -23,6 +24,7 @@ descriptions = OrderedDict([
 projects = dict(
     [("firedrake", "firedrakeproject"),
      ("PyOP2", "OP2"),
+     ("SLOPE", "coneoproject"),
      ("tsfc", "firedrakeproject"),
      ("COFFEE", "coneoproject"),
      ("ufl", "firedrakeproject"),
```
Comment (wence-) [inline scripts/firedrake-zenodo]: This barfs if the user calls `firedrake-zenodo` but has not installed SLOPE, which is an optional module.  Maybe something like:

```patch
diff --git a/scripts/firedrake-zenodo b/scripts/firedrake-zenodo
index 1c6c480..24a7ae7 100755
--- a/scripts/firedrake-zenodo
+++ b/scripts/firedrake-zenodo
@@ -34,6 +34,8 @@ projects = dict(
 
 components = descriptions.keys()
 
+optional_components = ("SLOPE", )
+
 parser = ArgumentParser(description="""Create Zenodo DOIs for specific versions of Firedrake components.
 
 If you are a Firedrake user, this script creates a JSON file encoding
@@ -135,8 +137,12 @@ def collect_repo_shas():
                 sys.exit(0)
             shas[component] = check_output(["git", "rev-parse", "HEAD"]).strip()
         except:
-            log.error("Failed to retrieve git hash for %s" % component)
-            raise
+            if component in optional_components:
+                log.warning("Failed to retrieve git hash for optional "
+                            "component '%s', continuing without it" % component)
+            else:
+                log.error("Failed to retrieve git hash for %s" % component)
+                raise
     return shas
 
 if args.release_tag:
```

Comment (FabioLuporini) [inline scripts/firedrake-zenodo]: My bad. I thought for a moment this was handled (so we are not zenodo-ing firedrake-adjoint? I only now notice it's not in the ``components`` dictionary). When I tried it, I was in the repo with SLOPE installed :)

Comment (FabioLuporini) [inline scripts/firedrake-zenodo]: I've added your patch and now seems to be OK

File: scripts/firedrake-install [hunk 2]
```diff
@@ -688,6 +695,31 @@ def build_and_install_adjoint():
...
+        check_call(["git", "clean", "-f", "-x", "-d"])
+        # Build and install (hack: need to do this manually as SLOPE is
+        # still stupid and doesn't have a proper configure+make+install system)
+        check_call(["make"])
+        check_call(["mv"] + glob(os.path.join('lib', '*')) + [os.path.join(sys.prefix, 'lib')])
+        include_dir = os.path.join(sys.prefix, 'include', 'SLOPE')
+        check_call(["mkdir", "-p", include_dir])
+        check_call(["cp"] + glob(os.path.join('sparsetiling', 'include', '*')) + [include_dir])
+        # Also install the Python interface
+        run_cmd(pyinstall)
```
Comment (wence-) [inline scripts/firedrake-install]: Can you use the "install" command instead?  That way, pyslope will be installed using pip and will pick up "developer" flags and similar.

**Comment by wence-** [comment]
Can you also add `--slope` to travis so that it is at least built (to check for problems).

File: scripts/firedrake-install (hunk 2 above)
Comment (FabioLuporini) [inline scripts/firedrake-install]: OK

**Comment by FabioLuporini** [comment]
I haven't really followed the discussion with ``drone``, but should I do something similar (i.e. adding --slope) with it too?

**Comment by wence-** [comment]
Yes, I think so.  Add it to the `firedrake-install` line in the normal way.  I'll have to regenerated the `.drone.sec` file, but can do that after you've pushed your change.

File: scripts/firedrake-install (hunk 2 above)
Comment (FabioLuporini) [inline scripts/firedrake-install]: I'm trying to do this, but now after the installation finishes the virtual env is screwed for mysterious reasons. If I try to run anything, I get import errors such as ``ImportError: No module named pyop2`` (similar for firedrake). ``pip list`` doesn't show any of the Firedrake modules. 

What I did was simply to replace ``run_cmd`` with install("SLOPE/") (now after ``os.chdir("..")``, instead of before)

Comment (FabioLuporini) [inline scripts/firedrake-install]: I showed this to @miklos1 , but we couldn't find an explanation

**Comment by FabioLuporini** [comment]
I've pushed the new version, but the problem above persists, at least on my machine. Now I'm trying again to reinstall everything

**Comment by wence-** [comment]
SLOPE doesn't build OOTB on MacOS because it unconditionally includes <omp.h> but the compilers don't by default come with that library installed.  See the compile log on travis.  Can you please ensure that this works?

**Comment by FabioLuporini** [comment]
Yes. 

However, the installation still screws the virtual env when passing --developer . 

Keep you posted

**Comment by FabioLuporini** [comment]
I think this can safely be merged even though SLOPE doesn't support developer mode for the reasons discussed above

**Comment by miklos1** [comment]
I still had some issues (on the PyOP2 side at least) the last time I checked this. Let me try again first...

**Comment by miklos1** [comment]
OK, so SLOPE installation seems to work, I only experience problems when updating PyOP2.

**Comment by FabioLuporini** [comment]
What kind of problems?

Maybe with an update you have to remove the old ``fusion.pyc`` file ?

**Comment by FabioLuporini** [comment]
Nevertheless, this can be merged
